### ISSUE TITLE
Residual connections in the decoder allow cheating?

### ISSUE THREAD

**Issue by JannisBush** [tensorflow/tensor2tensor#1617]
The masked-multi-head attention prevents cheating by masking out the tokens in the input sentence to the right of the current word, but the residual connections are unmasked and allow cheating? 

Say we have our input sequence x for the decoder and we apply the masked multi-head attention and get some representation y (Sublayer(x)) that has no access to the tokens to the right due to the masking (-inf in the softmax), but now we apply the Residual-Add & Norm step `LayerNorm(x+ Sublayer(x))`, only Sublayer(x) is masked, in x we still have information about words following our current word. In all decoder layers and sublayers there are residual connections, therefore some information from the original input sequence x can flow unmasked to the output and the network can cheat? 

Where am I missing out/misunderstanding something? 
Why can't the network cheat, despite these residual connections?

**Comment by lukaszkaiser** [comment]
When we add, like in x + Sublayer(x), we are adding position-wise, so nothing is leaked.
